### PR TITLE
Feature lwdev 7410 recovery api

### PR DESCRIPTION
**Pull request by gmaf**

### CODE REVIEW

File: Core/Core.csproj [hunk 1]
```diff
@@ -1,21 +1,22 @@
...
-    <TargetFramework>netcoreapp2.0</TargetFramework>
+    <TargetFramework>netcoreapp2.1</TargetFramework>
     <Version>1.0.0</Version>
   </PropertyGroup>
   <ItemGroup>
     <PackageReference Include="Lykke.Service.Affiliate.Client" Version="1.0.7" />
     <PackageReference Include="Lykke.Service.AssetDisclaimers.Client" Version="3.0.0" />
     <PackageReference Include="Lykke.Service.CandlesHistory.Client" Version="2.1.0" />
     <PackageReference Include="Lykke.Service.ClientDialogs.Client" Version="1.0.0" />
+    <PackageReference Include="Lykke.Service.ClientAccountRecovery.Client" Version="1.0.17-beta" />
```
Comment (gmaf) [inline Core/Core.csproj]: Package would be updated after stable nuget release of Lykke.Service.ClientAccountRecovery.

File: LkeServices/LkeServices.csproj [hunk 2]
```diff
@@ -9,13 +9,19 @@
...
   </ItemGroup>
   <ItemGroup>
+    <PackageReference Include="AutoMapper" Version="7.0.1" />
+    <PackageReference Include="jose-jwt" Version="2.4.0" />
     <PackageReference Include="Lykke.Payments.Client" Version="1.0.8" />
     <PackageReference Include="Lykke.Service.Balances.Client" Version="1.1.2" />
     <PackageReference Include="Lykke.Service.ClientAccount.Client" Version="1.3.0" />
     <PackageReference Include="Lykke.Service.Assets.Client" Version="4.0.1" />
-    <PackageReference Include="Lykke.Service.Session.Client" Version="1.20.1" />
+    <PackageReference Include="Lykke.Service.ClientAccountRecovery.Client" Version="1.0.17-beta" />
```
Comment (gmaf) [inline LkeServices/LkeServices.csproj]: Package would be updated after stable nuget release of Lykke.Service.ClientAccountRecovery.

File: LkeServices/LkeServices.csproj [hunk 3]
```diff
@@ -9,13 +9,19 @@
...
+    <PackageReference Include="Lykke.Service.PersonalData" Version="1.0.82" />
     <PackageReference Include="Lykke.Service.Kyc.Client" Version="1.0.220" />
     <PackageReference Include="Microsoft.AspNetCore.Http.Abstractions" Version="2.1.1" />
+    <PackageReference Include="Lykke.Service.Session.Client" Version="2.1.0-beta" />
```
Comment (gmaf) [inline LkeServices/LkeServices.csproj]: Package would be updated after stable nuget release of Lykke.Service.Session.

File: LykkeApi2/LykkeApi2.csproj [hunk 4]
```diff
@@ -39,29 +41,26 @@
     <PackageReference Include="Lykke.Service.OperationsRepository.Client" Version="1.0.97" />
     <PackageReference Include="Lykke.Service.OperationsRepository.Client.Core" Version="1.0.31" />
     <PackageReference Include="Lykke.Service.OperationsRepository.Contract" Version="1.0.13" />
-    <PackageReference Include="Lykke.Service.PersonalData" Version="1.0.79" />
+    <PackageReference Include="Lykke.Service.PersonalData" Version="1.0.82" />
     <PackageReference Include="Lykke.Service.PaymentSystem.Client" Version="1.0.22" />
     <PackageReference Include="Lykke.Service.RateCalculator.Client" Version="1.0.12" />
     <PackageReference Include="Lykke.Service.Registration" Version="2.1.2" />
-    <PackageReference Include="Lykke.Service.Session.Client" Version="1.20.1" />
+    <PackageReference Include="Lykke.Service.Session.Client" Version="2.1.0-beta" />
```
Comment (gmaf) [inline LykkeApi2/LykkeApi2.csproj]: Package would be updated after stable nuget release of Lykke.Service.Session.

File: LykkeApi2/Automapper/RecoveryAutomapperProfile.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+namespace LykkeApi2.Automapper
+{
+    internal class RecoveryAutomapperProfile : Profile
+    {
+        public RecoveryAutomapperProfile()
+        {
+            CreateMap<RecoveryStartRequestModel, RecoveryStartDto>()
+                .ForMember(dest => dest.Ip,
+                    opt => opt.ResolveUsing((src, dest, destMember, context) => context.Items["Ip"]))
```
Comment (gmaf) [inline LykkeApi2/Automapper/RecoveryAutomapperProfile.cs]: This solution is not final, I used Autofact for easier mappings between request models.
I don't like ```ResolveUsing``` and context dictionary for passing fields which are outside of mapping models.
for example ```context.Items["Ip"]```.
```cs
var startRecoveryDto = _mapper.Map<RecoveryStartRequestModel, RecoveryStartDto>(model,
    opt =>
    {
        opt.Items["Ip"] = _requestContext.GetIp();
        opt.Items["UserAgent"] = _requestContext.UserAgent;
    });
```


Direct mapping using Object initializer was used before. It looks cumbersome, but eliminates possible errors related to context dictionary key names.
Example:
```cs
var startRecoveryDto = new RecoveryStartDto
{
    Email = model.Email,
    PartnerId = model.PartnerId,
    Ip = _requestContext.GetIp(),
    UserAgent = _requestContext.UserAgent
};
```


If it's better to construct request models directly, without using Autofac, I could return back old direct mapping approach.

File: Core/Settings/APIv2Settings.cs [hunk 6]
```diff
@@ -161,5 +164,10 @@ public static string GetOrderBookKey(this CacheSettings settings, string assetPa
             return string.Format(settings.OrderBooksCacheKeyPattern, assetPairId, isBuy);
         }
     }
+
+    public class ClientAccountRecoverySettings
+    {
+        public long SelfieImageMaxSizeMBytes { get; set; }
```
Comment (MikhailGerasimov) [inline Core/Settings/APIv2Settings.cs]: Maybe it is better to make the property optional with a default value = 64Mb

File: Core/Domain/Recovery/RecoveryState.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
+﻿using Lykke.Service.ClientAccountRecovery.Client.AutoRestClient.Models;
```
Comment (MikhailGerasimov) [inline Core/Domain/Recovery/RecoveryState.cs]: Is the namespace correct?

File: Core/Domain/Recovery/RecoveryState.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+
+namespace Core.Domain.Recovery
+{
+    public class RecoveryState
```
Comment (MikhailGerasimov) [inline Core/Domain/Recovery/RecoveryState.cs]: Add a description for the classes in public API

File: LykkeApi2/Middleware/LykkeApiError/LykkeApiErrorMiddleware.cs [hunk 9]
```diff
@@ -1,45 +1,44 @@
...
+using Core.Exceptions;
+using JetBrains.Annotations;
+using LykkeApi2.Models;
+using Microsoft.AspNetCore.Http;
+using Newtonsoft.Json;
+
+namespace LykkeApi2.Middleware.LykkeApiError
+{
+    [UsedImplicitly]
+    public class LykkeApiErrorMiddleware
```
[the remark this answers is not in the document]
Comment (gmaf) [reply LykkeApi2/Middleware/LykkeApiError/LykkeApiErrorMiddleware.cs]: Yes, it is intended to be internal, thanks for pointing this out.

File: Core/Settings/APIv2Settings.cs (hunk 6 above)
Comment (gmaf) [reply Core/Settings/APIv2Settings.cs]: Yes it seems that namespace is correct. As we discussed to use enums directly from client library to bypass redundant mappings.
image

File: Core/Domain/Recovery/RecoveryState.cs (hunk 7 above)
Comment (gmaf) [reply Core/Domain/Recovery/RecoveryState.cs]: Yes it seems that namespace is correct. As we discussed to use enums directly from client library to bypass redundant mappings.
image

File: LkeServices/LkeServices.csproj (hunk 2 above)
Comment (gmaf) [reply LkeServices/LkeServices.csproj]: Updated

File: Core/Core.csproj (hunk 1 above)
Comment (gmaf) [reply Core/Core.csproj]: Updated

File: LykkeApi2/Models/Recovery/RecoveryCompleteRequestModel.cs [hunk 10]
```diff
@@ -0,0 +1,13 @@
+﻿namespace LykkeApi2.Models.Recovery
+{
+    public class RecoveryCompleteRequestModel
```
[the remark this answers is not in the document]
Comment (gmaf) [reply LykkeApi2/Models/Recovery/RecoveryCompleteRequestModel.cs]: Done.

File: LykkeApi2/Models/LykkeApiErrorResponse.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+﻿using Newtonsoft.Json;
+
+namespace LykkeApi2.Models
+{
+    public class LykkeApiErrorResponse
```
[the remark this answers is not in the document]
Comment (gmaf) [reply LykkeApi2/Models/LykkeApiErrorResponse.cs]: Done.

File: Core/Domain/Recovery/RecoveryState.cs (hunk 8 above)
Comment (gmaf) [reply Core/Domain/Recovery/RecoveryState.cs]: Done.

File: Core/Settings/APIv2Settings.cs (hunk 6 above)
Comment (gmaf) [reply Core/Settings/APIv2Settings.cs]: Done, default is 32Mb.

File: LykkeApi2/LykkeApi2.csproj [hunk 12]
```diff
@@ -39,29 +41,26 @@
...
-    <PackageReference Include="Microsoft.Extensions.Caching.Redis" Version="2.0.2" />
-    <PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="2.0.2" />
+    <PackageReference Include="Microsoft.AspNetCore" Version="2.1.2" />
+    <PackageReference Include="Microsoft.AspNetCore.Authentication.JwtBearer" Version="2.1.1" />
+    <PackageReference Include="Microsoft.AspNetCore.StaticFiles" Version="2.1.1" />
+    <PackageReference Include="Microsoft.Extensions.Caching.Redis" Version="2.1.1" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Debug" Version="2.1.1" />
     <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="2.1.1" />
     <PackageReference Include="Lykke.Service.Assets.Client" Version="4.0.1" />
     <PackageReference Include="Swashbuckle.AspNetCore.Swagger" Version="1.1.0" />
```
[the remark this answers is not in the document]
Comment (gmaf) [reply LykkeApi2/LykkeApi2.csproj]: I think it's better to remove these references in a separate PR.
This PR already contains too much logic for different tasks.

File: LykkeApi2/Automapper/RecoveryAutomapperProfile.cs (hunk 5 above)
Comment (gmaf) [reply LykkeApi2/Automapper/RecoveryAutomapperProfile.cs]: Thanks for comment.

File: Core/Exceptions/LykkeApiErrorException.cs [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+        /// <returns>
+        ///     New <see cref="LykkeApiErrorException" /> with <see cref="StatusCode" /> set to
+        ///     <see cref="HttpStatusCode.NotFound" />.
+        /// </returns>
+        public static LykkeApiErrorException NotFound(ILykkeApiErrorCode lykkeApiErrorCode,
+            string message = "")
+        {
+            return new LykkeApiErrorException(HttpStatusCode.NotFound, lykkeApiErrorCode, message);
+        }
+
```
[the remark this answers is not in the document]
Comment (gmaf) [reply Core/Exceptions/LykkeApiErrorException.cs]: The idea was to limit developers who use this exception with predefined constructors and http codes. If someone really needs one, he should add new constructor and use it.

File: LkeServices/LkeServices.csproj [hunk 14]
```diff
@@ -9,13 +9,19 @@
     <None Remove="Setthings\**" />
   </ItemGroup>
   <ItemGroup>
+    <PackageReference Include="AutoMapper" Version="7.0.1" />
+    <PackageReference Include="jose-jwt" Version="2.4.0" />
```
Comment (simpllfy) [inline LkeServices/LkeServices.csproj]: Looks like this package is not used.

File: LkeServices/LkeServices.csproj [hunk 15]
```diff
@@ -9,13 +9,19 @@
...
     <PackageReference Include="Lykke.Payments.Client" Version="1.0.8" />
     <PackageReference Include="Lykke.Service.Balances.Client" Version="1.1.2" />
     <PackageReference Include="Lykke.Service.ClientAccount.Client" Version="1.3.0" />
     <PackageReference Include="Lykke.Service.Assets.Client" Version="4.0.1" />
-    <PackageReference Include="Lykke.Service.Session.Client" Version="1.20.1" />
+    <PackageReference Include="Lykke.Service.ClientAccountRecovery.Client" Version="1.0.18" />
+    <PackageReference Include="Lykke.Service.PersonalData" Version="1.0.82" />
     <PackageReference Include="Lykke.Service.Kyc.Client" Version="1.0.220" />
     <PackageReference Include="Microsoft.AspNetCore.Http.Abstractions" Version="2.1.1" />
+    <PackageReference Include="Lykke.Service.Session.Client" Version="2.1.0-beta" />
```
[the remark this answers is not in the document]
Comment (simpllfy) [reply LkeServices/LkeServices.csproj]: So, let's do it?

**Pull request closed without merging**